libs/go-libs/sync/shared/sharedarr.go

```diff
@@ -0,0 +1,228 @@
+package shared
+
+import "sync"
+
+type SharedArr[T any] struct {
+	Shared[[]*Shared[T]]
+}
+
+func (s *SharedArr[T]) Add(new *Shared[T]) {
+	s.WLock()
+	defer s.WUnlock()
+
+	s.UnsafeAdd(new)
+}
+
+func (s *SharedArr[T]) UnsafeAdd(new *Shared[T]) {
+	idx := s.UnsafeFind(new)
+
+	if idx < 0 {
+		*s.Val = append(*s.Val, new)
+	}
+}
+
+func (s *SharedArr[T]) Remove(new *Shared[T]) *Shared[T] {
+	s.WLock()
+	defer s.WUnlock()
+
+	return s.UnsafeRemove(new)
+}
+
+func (s *SharedArr[T]) UnsafeRemove(new *Shared[T]) *Shared[T] {
+	idx := s.UnsafeFind(new)
+
+	if idx == -1 {
+		return nil
+	}
+
+	if idx == 0 {
+		*s.Val = (*s.Val)[1:]
```

Comment:
I think this will fail if the array has a size of 1 (I have a doubt)

Reply:
In fact it works with Size == 1 but it failed with Size == 0
But Size == 0 is not a possible case here.